src/lib/scan-imports.ts

```diff
@@ -0,0 +1,158 @@
+ */
+function importModuleName(node: NodePath<ImportDeclaration>): string | undefined {
+  if (isImportDeclaration(node)) {
+    return node.node.source.value;
+  }
+  return undefined;
+}
+
+/**
+ * Return dynamic import name (or undefined)
+ */
+function dynamicImportModuleName(node: NodePath<CallExpression>): string | undefined {
+  if (isCallExpression(node)) {
+    if (node.node.callee.type === 'Import' && node.node.arguments && node.node.arguments.length) {
+      return (node.node.arguments[0] as any).value;
+    }
+  }
+  return undefined;
+}
+
+/**
+ * Return array of dependencies
+ */
+function parseImports(file: string) {
+  const code = fs.readFileSync(file, 'utf8');
+
+  // skip file if empty, or if permissions error
+  if (!code) {
+    return [];
+  }
+  const deps = new Set<string>();
+
+  // try to parse; this will fail if not JS
+  try {
+    const ast = parse(code, {plugins: ['dynamicImport'], sourceType: 'module'});
+    traverse(ast, {
+      enter(node) {
+        const moduleName =
+          importModuleName(node as NodePath<ImportDeclaration>) ||
+          dynamicImportModuleName(node as NodePath<CallExpression>);
+
+        if (moduleName) {
+          deps.add(moduleName);
+        }
+      },
+    });
+    return [...deps];
+  } catch (e) {
+    return [];
+  }
+}
+
+interface ScanOptions {
+  dependencies: {[key: string]: string};
+  dest: string;
+}
+
+/**
+ * Scan glob pattern for imports
+ */
+export default function scanImports(entry: string, options: ScanOptions): string[] {
+  if (!Object.keys(options.dependencies).length) {
+    console.warn(`No dependencies or devDependencies found in package.json`);
+    return [];
+  }
+
+  const files = glob.sync(entry, {nodir: true});
+  if (!files.length) {
+    console.warn(`No files found matching ${entry}`);
+    return [];
+  }
+
+  // start perf benchmark
+  const spinner = ora(`Scanning ${entry}`).start();
+  const timeStart = process.hrtime();
+
+  // get all dependencies (even local ones)
+  const allDependencies = new Set<string>();
+  files.forEach(file => {
+    parseImports(file).forEach(dep => {
+      allDependencies.add(resolveWebModule(dep, options.dest));
+    });
+  });
+
+  // filter out modules not in options.dependencies and sort
+  const npmDependencies = [...allDependencies]
+    .map(originalName => {
+      // npm name (ex: `vue`)
+      if (!!options.dependencies[originalName]) {
+        return originalName;
+      }
+      // npm name + extension (ex: `graphql.js`)
+      const noExt = originalName.replace(/\.[^.]+$/, '');
+      if (!!options.dependencies[noExt]) {
+        return noExt;
+      }
+      // nested dep (ex: `algoliasearch/dist/algoliasearchLite.js`)
+      const [scope, name] = explode(originalName);
+      const rootName = scope.startsWith('@') ? `${scope}/${name}` : scope;
+      if (!!options.dependencies[rootName]) {
+        return originalName;
+      }
+      // dep not found
+      return undefined;
```

Comment:
This section is new, and handles the `algoliasearch/dist/algoliasearchLite.js` scenario (I was actually testing that scenario with `vue`, but you were right that I was bundling the wrong module).  This fixes that.

Hopefully with these comments, it’s clear that it’s only changing the `import` name in one scenario: if it finds a match in `package.json` when the file extension is removed (this is after `/web_modules` is stripped, so only things like `graphql.js` would be left.

Reply:
Nice, sounds good